core/src/main/scala/quasar/datasource/kafka/SeekConsumer.scala

```diff
@@ -79,54 +90,77 @@ class FullConsumer[F[_]: ConcurrentEffect: ContextShift: Timer, K, V](
    *
    * @param endOffsets Map of all _non-empty_ topic/partitions to their end offsets (offset of last message + 1)
    */
   def takeUntilEndOffsets(
-      // TODO: it would make more sense to take a KafkaConsumer, but that can't be mocked on fs2-kafka 1.0.0.
-      stream: Stream[F, Stream[F, CommittableConsumerRecord[F, K, V]]],
+      stream: Stream[F, Stream[F, CCR]],
       endOffsets: Map[TopicPartition, Long])
-      : Stream[F, CommittableConsumerRecord[F, K, V]] = {
+      : Stream[F, CCR] =
     stream
-      .take(endOffsets.size.toLong) // assumes no automatic assignment; needs a filter otherwise
-      .map(_.takeThrough(isNotOffsetLimit(_, endOffsets)))
+      .take(endOffsets.size.toLong)
+      .map(_.through(offsetPipe(endOffsets)))
       .parJoin(endOffsets.size)
-  }
 
-  /**
-   * True if there is at least one more record in the same topic/partition.
-   *
-   * Kafka API's `endOffsets` method returns the offset of the last available message plus one. Therefore,
-   * we return false if the current record corresponds to the offset of the last available message, or it
-   * is greater than that. The latter should not happen.
-   *
-   * @param committableRecord Last record read
-   * @param offsets map returned by Kafka's `endOffsets(Collection&lt;TopicPartition&gt;)` method
-   */
-  def isNotOffsetLimit(
-      committableRecord: CommittableConsumerRecord[F, K, V],
-      offsets: Map[TopicPartition, Long])
-      : Boolean = {
-    val record = committableRecord.record
-    val topic = record.topic
-    val partition = record.partition
-    val topicPartition = new TopicPartition(topic, partition)
-    val end = offsets(topicPartition)
-    log.trace(s"Read offset ${record.offset} / ${end - 1} from $topicPartition")
-    record.offset < (end - 1)  // From endOffsets javadoc: "the offset of the last available message + 1"
-  }
+  // Can't use `filter` since we need to stop stream from execution if no new records appeared
+  def offsetPipe(endOffsets: Map[TopicPartition, Long]): Pipe[F, CCR, CCR] =
+    stream => offsetPull(stream, endOffsets).stream
+
+  private def offsetPull(input: Stream[F, CCR], endOffsets: Map[TopicPartition, Long]): Pull[F, CCR, Unit] =
+    // if we `consumer.seek` to previous end and there is no new events, then this waits until new messages
+    // appear, that's why instead of `consumer.seek(p, w)` we do `consumer.seek(p, w - 1)`
+    input.pull.uncons1 flatMap {
+      case None =>
+        Pull.done
+      case Some((ccr, tail)) =>
+        val record = ccr.record
+        val topic = record.topic
+        val partition = record.partition
+        val topicPartition = new TopicPartition(topic, partition)
+        val previous = initialOffsets.get(partition).getOrElse(0L) - 1
+        // From endOffsets javadoc: "the offset of the last available message + 1"
+        val end = endOffsets(topicPartition) - 1
+        // No new events since previous push
+        if (end <= previous) {
```

Comment:
Is this something we could check earlier, when seeking? Or maybe using `consumer.partitionsMapStream`?

Reply:
`partitionsMapStream` is unavalable, we're tied to `1.1.x` version of fd4s-kafka unfortunately by custom `KafkaConsumer.java` (or maybe other `*.java` file). With `partitionsMapStream` we definitely could check that inner streams are unnecessary and don't do `+ 1` offset hack.